src/frequenz/sdk/timeseries/grid.py

```diff
@@ -78,7 +81,17 @@ def initialize(
         if grid_connections[0].metadata is None:
             raise RuntimeError("Grid metadata is None")
 
-        fuse = grid_connections[0].metadata.fuse
+        # The component graph does not store the metadata as a
+        # ComponentMetadata object but as a dict and the type-checker believes
+        # it is a ComponentMetadata object.
+        metadata = grid_connections[0].metadata
+
+        fuse: Fuse | None = None
+        if isinstance(metadata, ComponentMetadata):
+            fuse = metadata.fuse
+        elif isinstance(metadata, dict):
+            fuse_dict = metadata.get("fuse", None)
+            fuse = Fuse(**fuse_dict) if fuse_dict else None
```

Comment:
Also, can it really happen that it's either of both and not always a `dict` then? Why do we have the `if isinstance(metadata, ComponentMetadata)` at all?

Reply:
Previous to the latest suggestion from Sahas, the `initialize()` method accepted a `Iterable[Component]` type as input parameter. The data pipeline was passing the component retrieved from the Component Graph but in a test it was just passing `Components` without using the component graph. As I have explained in https://github.com/frequenz-floss/frequenz-sdk-python/pull/758/files#r1401933878, the component graph does not retrieve the right object type and the type-checker failed to validate this.

In the current version, we can remove checking for `ComponentMetadata` as we will be only getting the components from the Component Graph but this is hack caused by the Component Graph...ideally we should only be checking for `metadata is not None`

Reply:
yes, this is definitely wrong, the component graph shouldn't leak details about its internal implementation (of converting to dict and back), especially in this buggy fashion, that we have to deal with in other parts of the code.

Reply:
Alright, so I'd say removed the unnecessary check in this PR and create a new issue to fix the type-leakage in the component graph.

Reply:
Updated, the diff is here https://github.com/frequenz-floss/frequenz-sdk-python/compare/66abc038b0b9896a993c2b0b7000c43242526455..c10d6e2ced5491a112f29a9f6965de13e7b9b0ed